src/java_bytecode/java_types.cpp

```diff
@@ -167,6 +167,243 @@ exprt java_bytecode_promotion(const exprt &expr)
+  {
+    size_t start = i;
+    while(i < src.size())
+    {
+      // parameter is an object type or instantiated generic type
+      if(src[i] == 'L')
+      {
+        i = find_closing_semi_colon_for_reference_type(src, i);
+        break;
+      }
+
+      // parameter is an array
+      else if(src[i] == '[')
+        i++;
+
+      // parameter is a type variable
+      else if(src[i] == 'T')
+        i = src.find(';', i); // ends on ;
+
+      // type is an atomic type (just one character)
+      else
+      {
+        break;
+      }
+    }
+
+    std::string sub_str = src.substr(start, i - start + 1);
+    const typet &new_type = java_type_from_string(sub_str, class_name_prefix);
+    INVARIANT(new_type != nil_typet(), "Failed to parse type");
+
+    if(new_type.id() == ID_symbol)
+    {
+      UNREACHABLE;
+    }
+
+    type_list.push_back(new_type);
+  }
+  return type_list;
+}
+
+/// For parsing a class type reference
+/// \param src: The input string
+///   Either a signature: "Lpackage/class<TT;>.innerclass<Ljava/lang/Foo;>;
+///   Or a descriptor: "Lpackage.class$inner;
+/// \param class_name_prefix: The name of the class to use to prefix any found
+///   generic types
+/// \return The reference type if parsed correctly, no value if parsing fails.
+optionalt<reference_typet>
+build_class_name(const std::string src, const std::string &class_name_prefix)
+{
+  PRECONDITION(src[0] == 'L');
+  // ends on ;
+  if(src[src.size() - 1] != ';')
+    return optionalt<reference_typet>();
+
+  std::string container_class = gather_full_class_name(src);
+  std::string identifier = "java::" + container_class;
+  symbol_typet symbol_type(identifier);
+  symbol_type.set(ID_C_base_name, container_class);
+
+  std::size_t f_pos = src.find('<', 1);
+  if(f_pos != std::string::npos)
+  {
+    java_generic_typet result(symbol_type);
+    // get generic type information
+    do
+    {
+      std::size_t e_pos = find_closing_delimiter(src, f_pos, '<', '>');
+      if(e_pos == std::string::npos)
+        throw unsupported_java_class_signature_exceptiont("recursive generic");
+
+      add_generic_type_information(
+        result, src.substr(f_pos, e_pos - f_pos + 1), class_name_prefix);
+
+      // Look for the next generic type info (if it exists)
+      f_pos = src.find('<', e_pos + 1);
+
+    } while(f_pos != std::string::npos);
+    return result;
+  }
+
+  return java_reference_type(symbol_type);
+}
+
+/// Finds the closing semi-colon ending a ClassTypeSignature that starts at
+/// \p starting_point.
+/// \param src: The input string to work on.
+/// \param starting_point: The string position where the opening 'L' we want to
+///   find the closing ';' for.
+/// \return The string position corresponding to the matching ';'. For example:
+/// LA;, we would return 2. For LA<TT;>; we would return 7.
+/// See unit/java_bytecode/java_util_tests.cpp for more examples.
+size_t find_closing_semi_colon_for_reference_type(
```

Comment:
would it make sense to adapt the `find_closing_delimiter` function to take a set of opening characters? Then using `find_closing_delimiters(src, end, "LT", `;`)` would return the `;` that corresponds to the first `L` and skip over any other `L`,`;` or `T`, `;` pair

Reply:
I think the problem is you have two types of brackets `L`, `T`, which are closed by `;` and `<` which is closed by `>`. I suppose it would in fact work as `Lfoo<;>`; isn't legal, but presumably wouldn't bec caught if that did happen. Also `LL;` would fail (class named L)

Reply:
And indeed the far more common `Lgeneric_class<TT;>;`

Reply:
the idea of `find_closing_delimiters` is to support hierarchic structures (maybe it's buggy?) and allow for free definition of closing and opening delimiter.
What I imagined was to adapt it in such a way that it would in this case
 - ignore `<` and `>`
 - increase the depth at `T` and `L`
 - decrease depth at `;`

agreed for `LL;`  or `LT;`, but with one lookahead or a Boolean that indicates that the last char was an `L` and "opened" the object name, we would be good, wouldn't we?
This parsing is slightly fragile and with one common function, we'd at least know where to correct an error and don't risk overseeing another place.

Reply:
Surely if you ignore `<`, there is no difference between `LALFoo;` (a class called `ALFoo`) and `LA<LFoo;>;` and in the first case you won't find a matching `;`? 

I'll give it a go but I think this is a fudamentally different task (matching two isolated sets of brackets) and I think it's a bad idea to complicate `find_closing_delimiters`

Reply:
keep it like that then, with the overall goal to get in a real parser, I am ok with this